src/python/modules/TensorflowGraph/TensorflowGraphBA.py

```diff
@@ -0,0 +1,192 @@
+            self.r_err_operation = compute_reproj_err(
+                self.cam_holder,
+                self.x_holder,
+                self.w_holder,
+                self.feat_holder
+            )
+
+        dc, dx, dw = grad_tape.jacobian(
+            self.r_err_operation,
+            (
+                self.cam_holder,
+                self.x_holder,
+                self.w_holder
+            ),
+            experimental_use_pfor = False
+        )
+
+        self.r_err_grad_operation = flatten(
+            tf.concat(( dc, dx, tf.reshape(dw, [2, 1]) ), axis = 1),
+            column_major = True
+        )
+
+        self.w_err_grad_operation = grad_tape.gradient(
+            self.w_err_operation,
+            self.w_holder
+        )
+
+    def first_running(self):
+        '''Performs the first session running.'''
+
+        self.session.run(
+            (
+                self.w_err_operation,
+                self.r_err_operation
+            ),
+            feed_dict = self.get_feed_dict(0)
+        )
+
+        self.session.run(
+            (
+                self.w_err_grad_operation,
+                self.r_err_grad_operation
+            ),
+            feed_dict = self.get_feed_dict(0)
+        )
+
+    def output(self):
+        '''Returns calculation result.'''
+
+        return BAOutput(self.r_err, self.w_err, self.jacobian)
+
+    def get_feed_dict(self, i):
+        '''Returns feed dictionary for the needed jacobian part calculation.'''
+
+        return {
+            self.cam_holder: self.cams[self.obs[i, 0]],
+            self.x_holder: self.x[self.obs[i, 1]],
+            self.w_holder: self.w[i],
+            self.feat_holder: self.feats[i]
+        }
+
+    def calculate_objective(self, times):
+        '''Calculates objective function many times.'''
+
+        for _ in range(times):
+            # calculate reprojection and weight error part by part and
+            # then combine the parts together
+            result = tuple(
+                self.session.run(
+                    (
+                        self.r_err_operation,
+                        self.w_err_operation
+                    ),
+                    feed_dict = self.get_feed_dict(i)
+                )
+                for i in range(self.p)
+            )
+
+            result = zip(*result)
+            self.r_err = np.concatenate(result.__next__(), 0)
+            self.w_err = np.stack(result.__next__(), 0)
+
+    def calculate_jacobian(self, times):
+        ''' Calculates objective function jacobian many times.'''
+
+        for _ in range(times):
+            # calculate reprojection and weight error derivatives part by
+            # part. Note, that weight error should be added to the sparse
+            # matrix only after the last reprojection error jacobian part
+            # adding, otherwise the result will be wrong
+            dws = []
+            for i in range(self.p):
+                dr, dw = self.session.run(
```

Comment:
I guess there is a sparse matrix gather which might allow us to do this for all points at once?

That's fine as future work (i.e. create an issue) if so.

Reply:
The sparse matrix that is used now in all python midules was written by me, using C++ realization as a pattern. So, the current realization (of both python sparse matrix and C++ one) doesn't allow adding a batch of blocks. If you meant that calculating all jacobian blocks at once could be better, then I can say that it couldn't. Firstly, creating such a huge computational graph takes a lot of operating memory. Secondly, my experiments show, that the current realization works faster

Reply:
Thanks, great to know.